src/services/products.service.ts

```diff
@@ -3,66 +3,277 @@ import { UserType } from '../core/users/User';
+  backups: {
+    enabled: boolean;
+    sourceTierId?: string;
+  };
+  drive: {
+    tier: Tier;
+  };
+}
+
+interface AllVpnFeatures {
+  vpnOptions: Array<{
+    featureId: string;
+    sourceTierId: string;
+    tierLabel: string;
+  }>;
+}
+
 export class ProductsService {
   constructor(
     private readonly tiersService: TiersService,
     private readonly usersService: UsersService,
   ) {}
 
-  async getApplicableTierForUser({
+  private selectBestDriveTier(businessTiers: Tier[], individualTiers: Tier[]): Tier {
+    if (businessTiers.length > 0) {
+      return businessTiers.reduce((best, current) => {
+        const bestStorage = best.featuresPerService[Service.Drive].workspaces.maxSpaceBytesPerSeat;
+        const currentStorage = current.featuresPerService[Service.Drive].workspaces.maxSpaceBytesPerSeat;
+
+        const isCurrentTierBetter = currentStorage > bestStorage;
+        return isCurrentTierBetter ? current : best;
+      }, businessTiers[0]);
+    }
+
+    if (individualTiers.length > 0) {
+      return individualTiers.reduce((best, current) => {
+        const bestStorage = best.featuresPerService[Service.Drive].maxSpaceBytes;
+        const currentStorage = current.featuresPerService[Service.Drive].maxSpaceBytes;
+
+        const isCurrentTierBetter = currentStorage > bestStorage;
+        return isCurrentTierBetter ? current : best;
+      }, individualTiers[0]);
+    }
+
+    throw new TierNotFoundError('No drive tiers available');
+  }
+
+  private mergeNonDriveFeatures(tiers: Tier[], mergedFeatures: MergedTierFeatures): void {
+    for (const tier of tiers) {
+      this.mergeMailFeatures(tier, mergedFeatures);
+      this.mergeMeetFeatures(tier, mergedFeatures);
+      this.mergeVpnFeatures(tier, mergedFeatures);
+      this.mergeAntivirusFeatures(tier, mergedFeatures);
+      this.mergeBackupsFeatures(tier, mergedFeatures);
+    }
+  }
+
+  private mergeMailFeatures(tier: Tier, mergedFeatures: MergedTierFeatures): void {
+    const mailFeatures = tier.featuresPerService[Service.Mail];
+    if (mailFeatures.enabled && mailFeatures.addressesPerUser > mergedFeatures.mail.addressesPerUser) {
+      mergedFeatures.mail = {
+        enabled: true,
+        addressesPerUser: mailFeatures.addressesPerUser,
+        sourceTierId: tier.id,
+      };
+    }
+  }
+
+  private mergeMeetFeatures(tier: Tier, mergedFeatures: MergedTierFeatures): void {
+    const meetFeatures = tier.featuresPerService[Service.Meet];
+    if (meetFeatures.enabled && meetFeatures.paxPerCall > mergedFeatures.meet.paxPerCall) {
+      mergedFeatures.meet = {
+        enabled: true,
+        paxPerCall: meetFeatures.paxPerCall,
+        sourceTierId: tier.id,
+      };
+    }
+  }
+
+  private mergeVpnFeatures(tier: Tier, mergedFeatures: MergedTierFeatures): void {
+    const vpnFeatures = tier.featuresPerService[Service.Vpn];
+    // We cannot determine which featureId provides access to more zones
+    // maybe add this info to the tier object?
```

Comment:
To better manage VPN access per tier, we could include an `availableZones` field within the VPN object or something like that. What do you think @sg-gs ?